app/services/decisions/income_decision_tree.rb

```diff
@@ -317,10 +317,14 @@ def edit_dependants(add_blank: false)
       edit(:dependants)
     end
 
+    def after_dependants_finished
+      return edit(:partner_employment_status) if include_partner_in_means_assessment?
+
+      determine_showing_no_income_page
+    end
+
     def determine_showing_no_income_page
-      if include_partner_in_means_assessment?
-        edit(:partner_employment_status)
-      elsif crime_application.income&.all_income_over_zero? || crime_application.income&.client_self_employed?
+      if income&.all_income_over_zero? || income&.client_self_employed? || income&.partner_self_employed?
```

Comment:
Can we extract the logic in line 327 so that it can be used both here and in the IncomeAssessement::AnswersValidator